Proof code

Davidson-Souza opened the pull request:
This PR adds the logic for verifying proofs given the current Stump. A new `Proof` struct will hold all data needed for the process, this is very close to how the golang implementation handles it.

## Now, this code can
- Modify the current state, adding new items.
- Undo the current state
- Verify proofs against a given state

## Todos
- ~~Make tests from json like #1~~
- ~~Add more test cases~~
- Make the inner loops more idiomatic

Davidson-Souza commented:
Update: [9e8094e](https://github.com/mit-dci/rustreexo/pull/2/commits/9e8094e8344980e4b27c2a3721ad434c5f2eb4fa) adds some new tests, and also fixes some bugs I found.

kcalvinalvin commented on src/accumulator/proof.rs:
```diff
@@ -0,0 +1,293 @@
+        let mut proof_hashes_index = 0 as usize;
+
+        while row <= total_rows {
+            let res = util::detect_row_hashes(&targets, row, total_rows);
```
I *think* the loop here is incorrect because there's no sorted merge. Maybe I'm wrong but the rationale I had for sorting is something like this:

```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
00  01          04  05      
```

When you do

```
targets.push(next_to_prove);
nodes.push(HashAndPos::mask(hash, next_to_prove);
```

the nodes now become out of order. Say if we're proving `[00, 09, 11]` with the proof being `[01, 10]`:

`00` gets hashed with proof of `01` and the code would append `08` and `nodes` become: `[00, 09, 11, 08]`.
`09` wouldn't be able to hash up to `12` because the sibling check at line 169 would fail as `11` is not a sibling of `09`.

Go playground code: https://go.dev/play/p/W6ujymRI63G

kcalvinalvin replied:
> You mean that when adding a new target, I should sort the vector before continuing?

Basically yeah. The loop that I had in the Go code only works when the nodes are sorted. I think the bug only shows up when the leaves are not on the same row.

>  I think I should add more tests for this kind of tree, where leafs are not at the same row.

Yeah that'd be good. You could add `Roots` to the json testcases so that you don't have to generate the roots from the leaves during the tests. I don't think the current setup works since there's no deletions implemented yet.

kcalvinalvin reviewed (commented):
Some initial review. I believe `create_root_candidates` needs some changes to how it works.

Davidson-Souza commented:
@kcalvinalvin made a few changes based on your comments. I dropped  the `HashAndPos`, now it's just tuples. Nodes are also sorted, and added test case for this tree where nodes are not in the same row.

kcalvinalvin commented on src/accumulator/proof.rs:
```diff
@@ -0,0 +1,310 @@
+                    // If the current position is a root, we add that to our result and don't go any further
+                    if util::is_root_position(nodes[i].0, stump.leafs, total_rows) {
+                        calculated_root_hashes.push(nodes[i].1);
+                    } else if nodes.len() > i + 1 && (nodes[i].0 | 1 == nodes[i + 1].0) {
```
`nodes[i].0 | 1` is confusing for someone who's not familiar with the code. Could we make a `is_right_niece()` in utils and call that function?

kcalvinalvin commented on src/accumulator/proof.rs:
```diff
@@ -0,0 +1,310 @@
+
+                        Proof::sorted_push(&mut nodes, (next_to_prove, hash));
+
+                        i += 1;
```
Could we comment here that we're doing an additional `i += 1` because we've processed an extra node in this loop?

kcalvinalvin commented on src/accumulator/proof.rs:
```diff
@@ -0,0 +1,310 @@
+                        let hash = if util::is_left_niece(nodes[i].0) {
+                            types::parent_hash(&nodes[i].1, &self.hashes[proof_hashes_index])
+                        } else {
+                            types::parent_hash(&self.hashes[proof_hashes_index], &nodes[i].1)
```
`proof_hashes_index` is another Go hack since this is the best way to do iteration in Go. With Rust, you have `.iter()` on vectors so it'd be better have that instead of doing incrementing.

In the beginning of `create_root_hashes` function, we can do something like:

`let proof_hashes_iter = self.hashes.iter()`

and call `proof_hashes_iter.next()` in this branch. If the result is `None`, return an error.

kcalvinalvin commented on src/accumulator/proof.rs:
```diff
@@ -0,0 +1,310 @@
+        let mut proof_hashes_index = 0 as usize;
+
+        while row <= total_rows {
+            let res = util::detect_row_hashes(&targets, row, total_rows);
```
`detect_row_hashes()` will only work if `targets` are sorted. Since there's no requirement that the caller has to pass in the targets as sorted, we should be sorting the targets before we use them.

After we declare `nodes` we should be sorting both the nodes and the targets before using them.

```
      "leafs":[
         0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
        20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 
        40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59,
        60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 
        80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99
       ],
       "proof": [
         "e7cf46a078fed4fafd0b5e3aff144802b853f8ae459a4f0c14add3314b7cc3a6",
         "f299791cddd3d6664f6670842812ef6053eb6501bd6282a476bbbf3ee91e750c",
         "2f0fd1e89b8de1d57292742ec380ea47066e307ad645f5bc3adad8a06ff58608",
         "bb7208bc9b5d7c04f1236a82a0093a5e33f40423d5ba8d4266f7092c3ba43b62",
         "3e23e8160039594a33894f6564e1b1348bbd7a0088d42c4acb73eeaed59c009d",
         "cd9c77062a338e63a63ca623db438cb8676f15466641079ee61ec2dda98de796",
         "83b2d85c9ae3143b1b1598076c85b3830d7f35a6c67c6ad122df8ab6235c40a1",
         "60e6d84b9ad3355211af775aff3338289431744c79fc06f5e2c065cbdd690f79",
         "9a899064135333436906fe5b714ca017e61d9e952860dcda410dec4bd0fd59a5",
         "18b41f49a19d41613b4f717b3ef291e2cde3b963fcf60f9a371affd761f1b998",
         "46c5f8e5b7ccdb3eb2014f988c30a73ba8bacfccf282293acbe559e56d540a17",
         "e799acb98a071c4884707e4bc8c093ba22571c8d84cc0223ab0c2c9327313a5b",
         "6faf95884aaf779221fbe2f81d68bbac26a70751d384fa5831b2d6d0cf07293d",
         "a8fdcb2277476c5a76bf6a5e1d107f5c48ec59a7c813f6f9f1419a648e5568af",
         "b151a956139bb821d4effa34ea95c17560e0135d1e4661fc23cedc3af49dac42",
         "94b8dc9cecafe156b80e4e64414cf4204858f56713776c525c38fb14c2c5e622",
         "8ec4ba1ba4d169b8a1e5261408d9799185d984206b1d8779e4a5c1e431023fc1"
       ],
       "targets": [
         10, 19, 20, 30, 31, 32, 99 
       ],
       "expected": true
     },
```

When we change the targets to:

```
20, 19, 10, 30, 31, 32, 99
```

the test will fail.

Davidson-Souza replied:
I was working with the assumption that `targets` is sorted. Sorting `targets` also requires permutations on `hashes`, right? We assume that a node's hash is at the same index as the node itself.

kcalvinalvin replied:
Yes we expect the caller to give the `hashes` in the same permutation as the `targets`. When we sort the targets, we should be changing the ordering of the `hashes` as well.

Davidson-Souza replied:
Just extrapolated the iterator idea, I think the version in [4ffc6c1](https://github.com/mit-dci/rustreexo/pull/2/commits/4ffc6c1ff9f5324dad985c6e637f13500041b5f3) is way more readable and safe than before. It also sorts at the beginning, so I think we don't have this bug anymore.  

Unrelated: CI is working :tada: . I made a basic integration with rustfmt in my repo, will test it here.

kcalvinalvin commented on src/accumulator/proof.rs:
```diff
@@ -0,0 +1,308 @@
+        let mut number_matched_roots = 0;
+
+        for i in 0..stump_roots {
+            if calculated_roots[number_matched_roots] == stump.roots[stump.roots.len() - (i + 1)] {
```
There's an edge case here where if the caller passes in an empty proof, there's an out of bounds error.

```
    { 
      "leafs":[
        0, 1, 2, 3, 4, 5
      ],
      "proof": [
      ],
      "targets": [
      ],
      "expected": true
    }, 
```

It is possible to have an empty proof during ibd if there's an empty block. Could also be an attack vector.

kcalvinalvin reviewed (commented):
Left more feedback

Davidson-Souza commented:
I think [462d808](https://github.com/mit-dci/rustreexo/pull/2/commits/462d8087ef76812bd48609494fd4413d427ab8ef) did address all your comments. Also improved `Verify`'s loop, now we don't have any risk of out-of-band read.  
I was also thinking, it is common in Rust creating a custom `Error` enum with all possible errors, and returning this instead of a string, it's easier to handle since you can `match` errors out. Do you think it's worth to implement it?

kcalvinalvin commented on src/accumulator/proof.rs:
```diff
@@ -0,0 +1,313 @@
+        // An iterator over proof hashes
+        let mut hashes_iter = self.hashes.iter();
+
+        for row in 0..(total_rows + 1) {
```
I do think `for row in 0..=total_rows` is more readable.

Was this done for backwards compatibility reasons for rust version < 1.26? `rust-bitcoin`'s minimum version is 1.41.1 so the backwards compatibility problem wouldn't apply here.

kcalvinalvin commented:
Looks good overall! (besides the one comment I made). Do you wanna squash it?

> I was also thinking, it is common in Rust creating a custom `Error` enum with all possible errors, and returning this instead of a string, it's easier to handle since you can `match` errors out. Do you think it's worth to implement it?

I think it'd be good to look at it in the future when we have ways to recover from certain errors. I don't think it matters much now because you can't recover from the error even if you know what kind of error it is.

Davidson-Souza commented:
> I do think for row in 0..=total_rows is more readable.  

Yes, it looks better without this parenthesis. I changed this loop and inside verify, instead of `while let` I'm using `for´ over the iterator, I think it's more clear that way.

>  Do you wanna squash the commits for merging?  

Squashed out, with my last changes too.

> > I was also thinking, it is common in Rust creating a custom Error enum with all possible errors, and returning this instead of a string, it's easier to handle since you can match errors out. Do you think it's worth to implement it?  

> I think it'd be good to look at it in the future when we have ways to recover from certain errors. I don't think it matters much now because you can't recover from the error even if you know what kind of error it is.  

Yeah, it can be something we do in the future. I was just curious about your opinion on that.
